### ISSUE TITLE
opt: panic in (*Set).IsUnconstrained(...) via Examples-ORMS/TestGorm

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#37073]
Seen in https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&buildTypeId=Cockroach_ExampleORMs&buildId=1258741

```
[11:55:31]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
[11:55:31]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
[11:55:31]
	panic: panic while executing 1 statements: SELECT * FROM _ INNER JOIN _ ON _._ = _._ WHERE (_ != _); caused by runtime error: invalid memory address or nil pointer dereference
[11:55:31]
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb0 pc=0x1b06252]
[11:55:31]
[11:55:31]
goroutine 917 [running]:
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0057bad80, 0x3ae4fe0, 0xc0057c39c0, 0x2f74540, 0x580a0a0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0057bad80, 0x3ae4fe0, 0xc0057c39c0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:434 +0x61
[11:55:31]
panic(0x2f74540, 0x580a0a0)
[11:55:31]
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc0067b6868)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:138 +0x10c
[11:55:31]
panic(0x2f74540, 0x580a0a0)
[11:55:31]
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/constraint.(*Set).IsUnconstrained(...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/constraint/constraint_set.go:116
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*logicalPropsBuilder).buildFiltersItemProps(0xc0074f40f8, 0x582a160, 0x582a170)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:1125 +0xf2
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*FiltersItem).ScalarProps(0x582a160, 0xc0074f4000, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:6581 +0x65
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).IsContradiction(0xc0057bb8a8, 0x582a160, 0xc007500770)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/custom_funcs.go:416 +0x39
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc0057bb898, 0x3b1b720, 0xc007500570, 0x3b1b720, 0xc007500770, 0x582a160, 0x1, 0x1, 0xc0075009a8, 0x10, ...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1586 +0xf58
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructInnerJoin(0xc0057bb898, 0x3b1b720, 0xc007500570, 0x3b1b720, 0xc007500770, 0xc007069200, 0x2, 0x2, 0xc0075009a8, 0xc005602000, ...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1206 +0x1c4
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).DynamicConstruct(0xc0057bb898, 0xc0074f0049, 0xc0074f8740, 0x4, 0x4, 0x1, 0x1)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:16273 +0x3b63
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructSelect(0xc0057bb898, 0x3b1ac80, 0xc007500970, 0xc00752c1b0, 0x1, 0x1, 0x0, 0x3b04640)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:358 +0x1ddc
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildWhere(0xc006d41260, 0xc0074927a0, 0xc00741fe18)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:741 +0x23c
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelectClause(0xc006d41260, 0xc00746d290, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00741fb00, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:636 +0x9a
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc006d41260, 0xc007475540, 0x0, 0x0, 0x0, 0xc00741fb00, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:586 +0x201
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc006d41260, 0x3aea860, 0xc007475540, 0xc00741fb00, 0x4)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:211 +0x204
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc006d41260, 0x0, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:152 +0x111
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*optPlanningCtx).buildExecMemo(0xc0057bb838, 0x3ae50a0, 0xc007481890, 0xc0067b6a18, 0xe44943, 0x3ae50a0, 0xc007481890)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:404 +0x18e
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0057bb0d0, 0x3ae50a0, 0xc007481890, 0x33963b3, 0x19, 0x2c96645c02248973, 0x5cc04dfd)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:147 +0x72
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0057bad80, 0x3ae50a0, 0xc007481890, 0xc0057bb0d0, 0x6, 0x7159df)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:761 +0x12e
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0057bad80, 0x3ae50a0, 0xc007481890, 0xc0057bb0d0, 0x7fbad79dde78, 0xc000ad6e70, 0x0, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:648 +0x148
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0057bad80, 0x3ae50a0, 0xc007481890, 0x3aea860, 0xc007475540, 0xc00678a5f6, 0x76, 0x0, 0x0, 0x0, ...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:412 +0xafe
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0057bad80, 0x3ae50a0, 0xc007481890, 0x3aea860, 0xc007475540, 0xc00678a5f6, 0x76, 0x0, 0x0, 0x0, ...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:100 +0x610
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0057bad80, 0x3ae50a0, 0xc007481860, 0x0, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1200 +0x1acf
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0057bad80, 0x3ae4fe0, 0xc0057c39c0, 0xc000ab32f8, 0x5400, 0x15000, 0xc000ab3390, 0xc0066aa0b0, 0x0, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1136 +0x1b4
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000879080, 0x3ae4fe0, 0xc0057c39c0, 0xc0057bad80, 0x5400, 0x15000, 0xc000ab3390, 0xc0066aa0b0, 0x0, 0x0)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436 +0xce
[11:55:31]
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0066523af, 0xc005543480, 0x3ae4fe0, 0xc0057c39c0, 0xc0066aa0b0, 0xc000879080, 0xc006786000, 0x3ae8860, 0xc005543460, 0xc00554ac00, ...)
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21f
[11:55:31]
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
[11:55:31]
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:515 +0x17b
[11:55:31]

```

**Comment by RaduBerinde** [comment]
Query (from log) is `SELECT * FROM products INNER JOIN order_products ON order_products.product_id = products.id WHERE (1 != 1)`.

For one, we don't handle `FalseExpr` in `buildConstraints`. But AFAICT we should return `unconstrained`, not `nil`.

**Comment by RaduBerinde** [comment]
I bet this is an instance of #36148, where two threads are trying to build the constraints for `FalseSingleton` and stepping over each other's toes. We really need to stop lazy-building properties, it is also problematic with plan caching.

**Comment by RaduBerinde** [comment]
This particular crash was fixed by #37972. The larger issue is still open under #36148.
